### PR TITLE
besselj Float32 calling incorrect function

### PR DESCRIPTION
**Pull request by simonbyrne**
Hopefully should fix recent 32-bit build failures (e.g. #15259).

### CODE REVIEW

**Comment by tkelman** [comment]
Aha. Strange that this would be intermittent?

**Comment by simonbyrne** [comment]
Yeah, very odd. I don't know enough about the ABI, but I guess in certain cases we happened across the right bit pattern.

**Comment by yuyichao** [comment]
Looks like we only have very limitted test for `besselj` function, (only `0` for `Float32`?)

**Comment by simonbyrne** [comment]
I've added some more tests.

File: test/math.jl [hunk 1]
```diff
@@ -387,12 +387,20 @@ j33 = besselj(3,3.)
 @test besselj(-3,-3) == j33
 @test besselj(-3,3) == -j33
 @test besselj(3,-3) == -j33
+@test besselj(3,3f0) ≈ j33
```
Comment (tkelman) [inline test/math.jl]: do these have to be approximate because we can't reliably constrain different libm's accuracy here?

Comment (simonbyrne) [inline test/math.jl]: No, it's because they're calling different precision functions, so they won't be exactly equal.

Comment (tkelman) [inline test/math.jl]: but if we round the higher-precision value into the lower precision type can we reliably test exact equality?

**Comment by simonbyrne** [comment]
I think I broke travis by doing a bunch of force pushes, and so the most recent commit doesn't appear. Any way I can get it to restart?

File: test/math.jl (hunk 1 above)
Comment (simonbyrne) [inline test/math.jl]: In that case, it's up to the libm/openspecfun (there are 3 different C functions involved)

**Comment by tkelman** [comment]
https://github.com/JuliaLang/julia/commit/f0b8cc80f57694311b52c30900884329a7e6ed2f is merging the right commit, someone just clicked cancel...

**Comment by simonbyrne** [comment]
Isn't it 63da79bec808ea2b177dc34afc62ce9082de7a2f?

**Comment by tkelman** [comment]
Yeah, but f0b8cc8 is "Merge 63da79b into 7199602". PR builds run on the merge commit.

File: test/math.jl (hunk 1 above)
Comment (tkelman) [inline test/math.jl]: okay so if the answer is no even to the rounded version, then not worth worrying about

**Comment by simonbyrne** [comment]
ah, makes sense. I've restarted.

File: test/math.jl (hunk 1 above)
Comment (eschnett) [inline test/math.jl]: Why do all the new tests use an assignment instead of a comparison?

Comment (yuyichao) [inline test/math.jl]: it's `\approx` not `=`

Comment (eschnett) [inline test/math.jl]: Sorry. I need either new glasses or a new font.

Comment (tkelman) [inline test/math.jl]: and maybe it needs an even larger tolerance?

Comment (yuyichao) [inline test/math.jl]: Or maybe cast the RHS to `Float32`.

```jl
julia> 0.1320342f0 ≈ 0.1320341839246122
false

julia> 0.1320342f0 ≈ Float32(0.1320341839246122)
true
```

Comment (simonbyrne) [inline test/math.jl]: Hmm, perhaps we should change `isapprox` to use the larger of the 2 tolerances.

Comment (yuyichao) [inline test/math.jl]: That would make more sense. Probably a different issue/pr though...

Comment (tkelman) [inline test/math.jl]: Let's do `Float32` on the rhs now, and think about tweaking `isapprox` separately

**Comment by simonbyrne** [comment]
Okay, I've updated the `isapprox` checks.

**Comment by tkelman** [comment]
great, do merge when green

**Comment by tkelman** [comment]
appveyor failure https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.13888/job/jko2yp6908r11vb6 was #15294

**Pull request merged**